File: scripts/test_scaffold_fixture_case.py
#!/usr/bin/env python3
"""Regression tests for fixture case scaffolding."""

from __future__ import annotations

import sys
import tempfile
import unittest
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

import preflight
import scaffold_fixture_case


class ScaffoldFixtureCaseTests(unittest.TestCase):
    def test_scaffold_creates_case_index_and_valid_document_chain(self) -> None:
        with tempfile.TemporaryDirectory() as tmp:
            root = Path(tmp)

            scaffold_fixture_case.scaffold_fixture_case(
                root,
                feature="cache-fixture",
                doc_id="cache-refresh",
                title="缓存刷新",
                case_id="CASE-CACHE-999",
                source_type="release_regression",
                source_reference="coding-plugins cache refresh",
                optimization_target="验证缓存刷新链路闭环",
                covered_risk="缓存未刷新导致 Codex 使用旧链路",
                current_date="2026-07-02",
            )

            feature_root = root / "docs" / "coding-plugins" / "features" / "cache-fixture"
            self.assertTrue((root / "CASE-INDEX.md").exists())
            self.assertTrue((feature_root / "README.md").exists())
            self.assertTrue((feature_root / "requirements" / "cache-refresh-PRD.md").exists())
            self.assertTrue((feature_root / "technicals" / "cache-refresh-TDD.md").exists())
            self.assertTrue((feature_root / "technicals" / "cache-refresh-TID.md").exists())
            self.assertTrue((feature_root / "test-cases" / "cache-refresh-TCD.md").exists())
            self.assertTrue((feature_root / "plans" / "cache-refresh-IPD.md").exists())
            self.assertTrue((feature_root / "evidences" / "cache-refresh-TED.md").exists())

            case_index = (root / "CASE-INDEX.md").read_text(encoding="utf-8")
            self.assertIn("## cache-fixture", case_index)
            self.assertIn("case_id: CASE-CACHE-999", case_index)

            preflight.check_formal_fixture_case_index(root)
            preflight.check_feature_document_chain_closure(root)
            preflight.check_traceability_closure(root)

    def test_scaffold_rejects_nested_feature_or_doc_id(self) -> None:
        with tempfile.TemporaryDirectory() as tmp:
            root = Path(tmp)

            with self.assertRaises(ValueError):
                scaffold_fixture_case.scaffold_fixture_case(
                    root,
                    feature="area/cache",
                    doc_id="cache-refresh",
                    title="缓存刷新",
                    case_id="CASE-CACHE-999",
                    source_type="release_regression",
                    source_reference="coding-plugins cache refresh",
                    optimization_target="验证缓存刷新链路闭环",
                    covered_risk="缓存未刷新导致 Codex 使用旧链路",
                )

            with self.assertRaises(ValueError):
                scaffold_fixture_case.scaffold_fixture_case(
                    root,
                    feature="cache-fixture",
                    doc_id="cache/refresh",
                    title="缓存刷新",
                    case_id="CASE-CACHE-999",
                    source_type="release_regression",
                    source_reference="coding-plugins cache refresh",
                    optimization_target="验证缓存刷新链路闭环",
                    covered_risk="缓存未刷新导致 Codex 使用旧链路",
                )


if __name__ == "__main__":
    unittest.main()

File: tests/ts/test_scaffold_fixture_case.mjs
import assert from "node:assert/strict";
import { existsSync, mkdtempSync, readFileSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import test from "node:test";

import { scaffoldFixtureCase } from "../../src/lib/scaffold-fixture-case.ts";
import { inspectDocumentChain } from "../../src/lib/workflow-state.ts";

function scaffold(root, overrides = {}) {
  return scaffoldFixtureCase(root, {
    feature: "cache-fixture",
    docId: "cache-refresh",
    title: "缓存刷新",
    caseId: "CASE-CACHE-999",
    sourceType: "release_regression",
    sourceReference: "coding-plugins cache refresh",
    optimizationTarget: "验证缓存刷新链路闭环",
    coveredRisk: "缓存未刷新导致 Codex 使用旧链路",
    currentDate: "2026-07-02",
    ...overrides,
  });
}

test("TypeScript scaffold creates case index and valid document chain", () => {
  const root = mkdtempSync(join(tmpdir(), "coding-plugins-scaffold-fixture-"));
  try {
    scaffold(root);
    const featureRoot = join(root, "docs/coding-plugins/features/cache-fixture");

    for (const path of [
      "CASE-INDEX.md",
      "docs/coding-plugins/features/cache-fixture/README.md",
      "docs/coding-plugins/features/cache-fixture/requirements/cache-refresh-PRD.md",
      "docs/coding-plugins/features/cache-fixture/technicals/cache-refresh-TSD.md",
      "docs/coding-plugins/features/cache-fixture/test-cases/cache-refresh-TVD.md",
      "docs/coding-plugins/features/cache-fixture/plans/cache-refresh-TED.md",
      "docs/coding-plugins/features/cache-fixture/evidences/cache-refresh-VED.md",
    ]) {
      assert.equal(existsSync(join(root, path)), true, `missing ${path}`);
    }
    assert.equal(existsSync(join(root, "docs/coding-plugins/features/cache-fixture/technicals/cache-refresh-TDD.md")), false);
    assert.equal(existsSync(join(root, "docs/coding-plugins/features/cache-fixture/technicals/cache-refresh-TID.md")), false);
    assert.equal(existsSync(join(root, "docs/coding-plugins/features/cache-fixture/test-cases/cache-refresh-TCD.md")), false);
    assert.equal(existsSync(join(root, "docs/coding-plugins/features/cache-fixture/plans/cache-refresh-IPD.md")), false);
    assert.equal(existsSync(join(root, "docs/coding-plugins/features/cache-fixture/evidences/cache-refresh-TED.md")), false);

    const caseIndex = readFileSync(join(root, "CASE-INDEX.md"), "utf8");
    assert.ok(caseIndex.includes("## cache-fixture"));
    assert.ok(caseIndex.includes("case_id: CASE-CACHE-999"));

    const tedText = readFileSync(join(featureRoot, "plans/cache-refresh-TED.md"), "utf8");
    assert.ok(tedText.includes("source_hash: sha256:"));
    assert.ok(tedText.includes("related_docs:"));
    assert.equal(tedText.includes("related_specs:"), false);
    assert.equal(tedText.includes("related_technical:"), false);
    assert.ok(tedText.includes("## 执行锁定区"));
    assert.ok(tedText.includes("## 执行简报"));
    assert.ok(tedText.includes("## 任务总览"));
    assert.ok(tedText.includes("## 缓存刷新（TASK-001 / REQ-001）"));

    const state = inspectDocumentChain(root, { feature: "cache-fixture", docId: "cache-refresh" });
    assert.equal(state.state, "ready-for-execution");
  } finally {
    rmSync(root, { recursive: true, force: true });
  }
});

test("TypeScript scaffold rejects nested feature or doc id", () => {
  const root = mkdtempSync(join(tmpdir(), "coding-plugins-scaffold-fixture-invalid-"));
  try {
    assert.throws(() => scaffold(root, { feature: "area/cache" }), /feature must be a flat slug/);
    assert.throws(() => scaffold(root, { docId: "cache/refresh" }), /doc_id must be a flat slug/);
  } finally {
    rmSync(root, { recursive: true, force: true });
  }
});
